### PR TITLE
Initial Gemification

### PR DESCRIPTION
**Pull request by mamodom**

### CODE REVIEW

**Comment by RoryO** [comment]
Are you sure this layout is right so when it's added to a rails project the rake tasks and generators are found?

**Comment by mamodom** [comment]
Yes, I created a dummy app to test the gem

File: lib/generators/grnds/onetime/install/install_generator.rb [hunk 1]
```diff
@@ -0,0 +1,11 @@
+require 'rails/generators'
+
+module Grnds
+  module Onetime
+    class InstallGenerator < Rails::Generators::Base
+      def add_migration
```
Comment (RoryO) [inline lib/generators/grnds/onetime/install/install_generator.rb]: I had it in mind to not have to rely on migrations to create the table, but instead check if the table exists first and use the rake task to execute sql to create the table. Internally the Migrator does the same thing.

File: lib/tasks/script.rake [hunk 2]
```diff
@@ -0,0 +1,7 @@
+namespace :script do
+  desc 'Run all pending onetime scripts'
+  task run_pending_onetime: :environment  do
+    runner = Grnds::Onetime::Runner.new
+    runner.run_all
+  end
```
Comment (RoryO) [inline lib/tasks/script.rake]: There should be another task here like check_table_exists and then this is `task run_pending_onetime: :check_table_exists` and then `task check_table_exists: :environment`

File: lib/grnds/onetime/runner.rb [hunk 3]
```diff
@@ -0,0 +1,133 @@
+module Grnds
+  module Onetime
+    class Runner
+      module ChangeRegistry
```
Comment (RoryO) [inline lib/grnds/onetime/runner.rb]: this should be in lib/grnds/onetime/runner/change_registry.rb

**Comment by RoryO** [comment]
There's a couple tasks that need done to make gem isolated completely. The biggest is getting away from a generated migration to creating the check and create the table on the rake task

File: lib/tasks/onetime.rake [hunk 4]
```diff
@@ -0,0 +1,37 @@
+namespace :onetime do
+  def table_exists?
+    ActiveRecord::Base.connection.table_exists? 'onetime_scripts'
+  end
+
+  desc 'Checks if the required table exists'
+  task check_table_exists: :environment  do
+    if table_exists?
+      puts "The table onetime_scripts exists"
```
Comment (RoryO) [inline lib/tasks/onetime.rake]: No need for this

File: lib/tasks/onetime.rake [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    else
+      raise "The table onetime_scripts doesn't exist\nTo create the table run 'rake onetime:install'\n\n"
```
Comment (RoryO) [inline lib/tasks/onetime.rake]: I was hoping to do this automatically. You never have to think about creating the schema_migrations table when running rake db:migrate, it just checks to see if it exists and creates it if need be

File: lib/generators/grnds/onetime/onetime_generator.rb [hunk 6]
```diff
@@ -6,11 +6,19 @@ class OnetimeGenerator < Rails::Generators::NamedBase
...
       @now = Time.now
+
       template 'onetime_script.rb.erb', "script/onetime/#{stringified_timestamp}_#{file_name}.rb"
     end
 
     def stringified_timestamp
       @now.strftime '%Y%m%d%H%M%S'
     end
+
+    def owner_body
```
Comment (RoryO) [inline lib/generators/grnds/onetime/onetime_generator.rb:17]: Interesting idea. A bit up in the air about it...

Comment (mamodom) [inline lib/generators/grnds/onetime/onetime_generator.rb:17]: I noticed that bundler does that when running `bundle gem gem_name`

Comment (RoryO) [inline lib/generators/grnds/onetime/onetime_generator.rb:17]: Yeah a number of tools do it, so I think it's okay.

**Pull request closed without merging**